### ISSUE TITLE
Define / improve error handling strategy

### ISSUE THREAD

**Issue by mmartinbar** [witnet/witnet-rust#72]
The error handling strategy is yet to be properly defined across the different components in the witnet-rust node. Several tasks need to be studied and accomplished regarding this error handling / graceful stop:

In main thread (`core/src/actors/node.rs`):
- [ ] Actix system needs to be stopped properly
- [ ] Decisions shall be made about what to do when different type of signals are received by the application

In session actor (`core/src/actors/session.rs`):
- [ ] Decisions shall be made about what to do with the actor when it could not be registered. Currently it is being stopped but a full stop is not the right thing to do as a full stop will try to unregister the session.

In connections manager actor (`core/src/actors/connections_manager.rs`):
- [ ] Decisions shall be made about what to do with the actor when the server could not be started. For the time being, if that happens, the application panics.

**Comment by anler** [comment]
I'm going to comment here my findings so far.

I'll divide my comments in:

1. Different scenarios for error handling
1. A proposal for our implementation
1. Info about Actix components related to error handling

## 1. Different scenarios for error handling

Here I'll expose the different error-handling scenarios that I can think of:

1. Panic inside actors
1. Error inside an actor
1. Error inside an actor that can be restarted
1. Error inside an actor that needs to notify its manager
1. Error inside an actor that needs to stop the system

### Panic inside actors

This scenario is the most destructive one since panics are irrecoverable errors, we should try to avoid these situations altogether, but if they cannot be dodged then we have two options:

- The panic should take the system down: This is the default behaviour for actors scheduled in the system arbiter.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;

fn main() {
    exit(System::run(|| {
        let addr = MyActor.start();
        // Panic in actor will shut down the system
        addr.do_send(Panic);
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;
}

#[derive(Message)]
struct Panic;

impl Handler<Panic> for MyActor {
    type Result = ();

    fn handle(&mut self, _: Panic, _: &mut Self::Context) {
        panic!("MyActor panic");
    }
}
```

- The panic should NOT take the system down: In this case before running the actor we need to create a new arbiter with the `stop_system_on_panic` set to `false` and execute the actor that might panic in that arbiter.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;

fn main() {
    exit(System::run(|| {
        // Panic in actor won't shut down the system
        let addr = Arbiter::builder()
            .stop_system_on_panic(false)
            .start(|_| MyActor);
        addr.do_send(Panic);
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;
}

#[derive(Message)]
struct Panic;

impl Handler<Panic> for MyActor {
    type Result = ();

    fn handle(&mut self, _: Panic, _: &mut Self::Context) {
        panic!("MyActor panic");
    }
}
```

So, to summarize:
- we need to avoid panics whenever possible
- if panics can't be avoided, we need to decide if it takes the system down with it or not

### Error inside an actor

This is the simplest scenario, if we have an actor A, and there's an error condition inside it that prevents it from running, then the action to take is to just stop the actor via:

- `ctx.stop()`: Stop processing incoming messages and switch to a stopping state. The actor could restore it's state to running because its `stopping` method is called.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;

fn main() {
    exit(System::run(|| {
        let addr = MyActor.start();
        addr.do_send(Stop);
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;

    fn stopping(&mut self, _: &mut Self::Context) -> Running {
        println!("MyActor stopping");
        Running::Stop
    }
}

#[derive(Message)]
struct Stop;

impl Handler<Stop> for MyActor {
    type Result = ();

    fn handle(&mut self, _: Stop, ctx: &mut Self::Context) {
        // MyActor's stopping method called
        ctx.stop();
    }
}
```

- `ctx.terminate()`: Switch to a stopped state. No `stopping` method is called.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;

fn main() {
    exit(System::run(|| {
        let addr = MyActor.start();
        addr.do_send(Stop);
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;

    fn stopping(&mut self, _: &mut Self::Context) -> Running {
        println!("MyActor stopping");
        Running::Stop
    }
}

#[derive(Message)]
struct Stop;

impl Handler<Stop> for MyActor {
    type Result = ();

    fn handle(&mut self, _: Stop, ctx: &mut Self::Context) {
        // MyActor's stopping method not called
        ctx.terminate();
    }
}
```

### Error inside an actor that can be restarted

This is the only scenario where I see the use for a supervised actor (other scenarios exist but they require keeping a persistent state of how many times the actor has been restarted and it requires more work).

A supervised actor is an actor implementing the `Supervised` trait and one that is restarted automatically whenever it stops via `ctx.stop()` or `ctx.terminate()`.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;
use std::sync::{Arc, Mutex};

fn main() {
    // Needed to avoid dropping all the references to the
    // actor, in which case it won't be restarted
    let addr = Arc::new(Mutex::new(None));
    let addr_copy = addr.clone();
    exit(System::run(move || {
        let addr = Supervisor::start(|_| MyActor);
        addr.do_send(Stop);
        *addr_copy.lock().unwrap() = Some(addr);
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;

    fn stopped(&mut self, _: &mut Self::Context) {
        println!("MyActor stopped");
    }
}

impl Supervised for MyActor {
    fn restarting(&mut self, _: &mut Self::Context) {
        println!("MyActor restarting");
    }
}

#[derive(Message)]
struct Stop;

impl Handler<Stop> for MyActor {
    type Result = ();

    fn handle(&mut self, _: Stop, ctx: &mut Self::Context) {
        ctx.stop();
    }
}
```

So, IMO we should use a supervised actor whenever we have, say, an actor A that performs a periodic work, and for some reason its state gets tainted, in that case it should stop itself and since its supervised it will be restarted with a clean state.

### Error inside an actor that needs to notify its manager

This pattern is for actors that manage other actors, for example, say an actor A creates an actor B, and B needs to be stopped because of an error, but before stopping it should notify A about it.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;
use std::time::Duration;

fn main() {
    exit(System::run(move || {
        ManagerActor.start();
    }));
}

struct ManagerActor;
struct MyActor {
    manager: Addr<ManagerActor>,
}

#[derive(Message)]
struct ManagedStopped;

impl Actor for ManagerActor {
    type Context = Context<Self>;

    fn started(&mut self, ctx: &mut Self::Context) {
        let act = MyActor {
            manager: ctx.address(),
        };
        act.start();
    }
}

impl Handler<ManagedStopped> for ManagerActor {
    type Result = ();

    fn handle(&mut self, _: ManagedStopped, _: &mut Self::Context) {
        println!("ManagerActor received a ManagedStopped from MyActor");
    }
}

impl Actor for MyActor {
    type Context = Context<Self>;

    fn started(&mut self, ctx: &mut Self::Context) {
        ctx.run_later(Duration::from_secs(2), |_, ctx| {
            ctx.stop();
        });
    }

    fn stopped(&mut self, _: &mut Self::Context) {
        self.manager.do_send(ManagedStopped);
    }
}
```

### Error inside an actor that needs to stop the system

This is the scenario for actors that need to take the system down whenever they find an error. The solution is just to call `System::current().stop_with_code()` or `System::current().stop()` which is equivalent to `System::current().stop_with_code(0)`.

``` rust
extern crate actix;

use actix::prelude::*;
use std::process::exit;
use std::time::Duration;

fn main() {
    exit(System::run(move || {
        MyActor.start();
    }));
}

struct MyActor;

impl Actor for MyActor {
    type Context = Context<Self>;

    fn started(&mut self, ctx: &mut Self::Context) {
        ctx.run_later(Duration::from_secs(3), |_, _| {
            System::current().stop_with_code(123);
        });
    }
}
```

## 2. Proposals to apply to our implementation

> In main thread (`core/src/actors/node.rs`):
> - Actix system needs to be stopped properly
> - Decisions shall be made about what to do when different type of signals are received by the application

- Even if there's a panic, Actix will shut down the system correctly, all we need to do is write our actors in a way they perform cleanup before/after stopping.
- If we want to shut down the system ourselves, we use `System::current().stop()`.
- We should use `process::exit()` to return the appropriate error code back to the OS.
- Probably the only actors that should stop the system if they find an error are our manager actors.

> In session actor (`core/src/actors/session.rs`):
> - [ ] Decisions shall be made about what to do with the actor when it could not be registered. Currently it is being stopped but a full stop is not the right thing to do as a full stop will try to unregister the session.

- If it cannot respond to a request due to an error, but doing so doesn't affect the reception of other requests, then it should just log the error and/or return a response to the client letting it know that it couldn't fullfill its request
- On the other hand, if the error affects the reception of other requests, say it breaks the handshake, then the actor should stop (close the session) and notify the sessions manager

> In connections manager actor (`core/src/actors/connections_manager.rs`):
> - [ ] Decisions shall be made about what to do with the actor when the server could not be started. For the time being, if that happens, the application panics.

As probably should be the case in all the managers, if an actor that is crucial for the application to work finds an error and needs to stop, then it should stop the entire system/application, but as I said we should avoid panicking whenever possible, and use `System::current().stop_with_code()` instead.


## 3. Info about Actix components related to error handling

Here's some information about some Actix components that play a part in how actors and the system behave in the face of errors and panics.

### System

- This is the actor that manages the runtime
- A system is bound the the thread where is created, and can be accessed with `System::current()`
- A system can be stopped with `system.stop_with_code(<i32>)` or `system.stop()` (equivalent to `system.stop_with_code(0)`)
- Stopping the system means that all arbiters in the system are stopped, so is its event loop

### Arbiter

- Used to run tasks in the event loop
- They belong to a specific system
- Any `panic!` in an actor running inside the System arbiter will cause the System to shut down
- Any `panic!` inside an actor running inside a manually created Arbiter will NOT cause the System to shut down, unless the option `stop_system_on_panic` is set

### Supervisor

- Actors may implement the `Supervised` trait
- If a supervised actor stops or terminates by calling itself with `ctx.stop()` or `ctx.terminate()`, it is restarted with a new execution context

**Comment by aesedepece** [comment]
> In session actor (`core/src/actors/session.rs`):
>
> [...]
>
> If it cannot respond to a request due to an error, but doing so doesn't affect the reception of other requests, then it should just log the error and/or return a response to the client letting it know that it couldn't fullfill its request

I mostly agree, but I think no failure response should be sent back. That's how many other blockchains' P2P networks behave and how the handshake is implemented in our case.

> On the other hand, if the error affects the reception of other requests, say it breaks the handshake, then the actor should stop (close the session) and notify the sessions manager

This is totally true, but with a nuance: in production, session closing as punishment for network protocol misbehavior should always be delayed for some seconds. This strategy, similar to [_"tarpitting"_][tarpit] found in firewalls helps preventing some DoS attack vectors as the attacker can't tell punishment from latency.

[tarpit]: https://en.wikipedia.org/wiki/Tarpit_(networking)

**Comment by aesedepece** [comment]
IMO:
- The `Supervised` trait makes a lot of sense for "workers", that is, actors spawned for fulfilling a particular task and then just die. If they fail to perform the task, there's no doubt the situation should be handled by some coordinator that may want to reassign the task to a new worker.
- A supervisor actor should be an `Arbiter` unless the supervisor is itself supervised by other `Arbiter`. _"Every supervisor is an `Arbiter`"_ is probably the safest way to go so any panics or errors only impact the minimum number of branches in our supervision tree.
- All managers should have their own `Arbiters` so their uptime is guaranteed for the entire lifespan of the app.
- `ConfigurationManager` is probably one exception to the previous point. If configuration reading / validation fails, we do want the entire system to fall down. 
- `ConnectionsManager` and `JsonRpcServer` are also special. They have a warm-up stage in which they start TCP servers. A failure in that stage (probably caused by not being able to bind to an address or port) should take everything down. However, any network errors in runtime should be handled gracefully and if any actor dies the problem should never bubble up to the system.

**Comment by anler** [comment]
> The Supervised trait makes a lot of sense for "workers", that is, actors spawned for fulfilling a particular task and then just die. If they fail to perform the task, there's no doubt the situation should be handled by some coordinator that may want to reassign the task to a new worker.

Bear in mind that, if a supervised actor stops explicitly, it will be restarted, always, and automatically. So we either have this or we have a coordinator that performs the supervision.

> A supervisor actor should be an Arbiter unless the supervisor is itself supervised by other Arbiter. "Every supervisor is an Arbiter" is probably the safest way to go so any panics or errors only impact the minimum number of branches in our supervision tree.

This makes sense in Akka, but for what I've seen Actix is different, there's no: if an actor dies its supervisor is notified, because in actix the supervisor is the actor itself and all it does is restart with a new execution context. So I would rephrase as: If we need a supervisor, then we probably need a coordinator actor (I'm trying to avoid `Arbiter` since in Actix means another thing) instead of a `Supervised` actor.

> All managers should have their own Arbiters so their uptime is guaranteed for the entire lifespan of the app.

If this is the case then no actor can interact directly with the managers, but through their coordinators. Also probably instead of each manager having a coordinator we just need one coordinator that knows when a manager stops and restarts it. On the other hand if the coordinator stops, the system should stop.

> ConfigurationManager is probably one exception to the previous point. If configuration reading / validation fails, we do want the entire system to fall down.

Agree

> ConnectionsManager and JsonRpcServer are also special. They have a warm-up stage in which they start TCP servers. A failure in that stage (probably caused by not being able to bind to an address or port) should take everything down. However, any network errors in runtime should be handled gracefully and if any actor dies the problem should never bubble up to the system.

If we follow the pattern: coordinator in front of managers, then for actors like this we at least can apply the Circuit Breaker pattern and if, after some retries these actors cannot start, stop the entire system.

**Comment by anler** [comment]
Next steps for this issue (pointed by @aesedepece):

> we shall draft a coordination/supervision/arbitration tree for the entire app in the whiteboard
it could be a great resource for wrapping our minds about this but also to let everyone see our actors system at a glance

**Comment by anler** [comment]
Good point @mmartinbar ! Indeed tat's the right behavior and it's clearly stated in Actix documentation:

> Supervisor does not re-create actor, it just calls `restarting()`

Sorry for describing it wrong :pray:
